x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformScheduler.java

```diff
@@ -0,0 +1,388 @@
+                .append(",failureCount=")
+                .append(failureCount)
+                .append(",nextScheduledTimeMillis=")
+                .append(nextScheduledTimeMillis)
+                .append(",listener=")
+                .append(listener)
+                .append("]")
+                .toString();
+        }
+    }
+
+    private static final Logger logger = LogManager.getLogger(TransformScheduler.class);
+
+    private final Clock clock;
+    private final ThreadPool threadPool;
+    private final TimeValue schedulerFrequency;
+    private final ConcurrentPriorityQueue<ScheduledTransformTask> scheduledTasks;
+    /**
+     * Prevents two concurrent invocations of the "processScheduledTasks" method.
+     *
+     * Set to {@code true} before processing
+     * Set to {@code false} after processing (doesn't matter whether successful or not).
+     */
+    private final AtomicBoolean isProcessingActive;
+
+    public TransformScheduler(Clock clock, ThreadPool threadPool, Settings settings) {
+        this.clock = Objects.requireNonNull(clock);
+        this.threadPool = Objects.requireNonNull(threadPool);
+        this.schedulerFrequency = Transform.SCHEDULER_FREQUENCY.get(settings);
+        this.scheduledTasks = new ConcurrentPriorityQueue<>(
+            ScheduledTransformTask::getTransformId,
+            ScheduledTransformTask::getNextScheduledTimeMillis
+        );
+        this.isProcessingActive = new AtomicBoolean();
+    }
+
+    /**
+     * Starts the processing.
+     * Method {@code processScheduledTasks} is invoked according to its configured {@code schedulerFrequency}.
+     */
+    public void start() {
+        threadPool.scheduleWithFixedDelay(this::processScheduledTasks, schedulerFrequency, ThreadPool.Names.GENERIC);
+    }
+
+    // Visible for testing
+    void processScheduledTasks() {
+        // Prevent two concurrent invocations of the "processScheduledTasks" method
+        if (isProcessingActive.compareAndSet(false, true) == false) {
+            return;
+        }
+        logger.trace("Processing scheduled tasks started");
+        Instant processingStarted = clock.instant();
+        try {
+            processScheduledTasksInternal();
+        } finally {
+            // Make sure we clear the "isProcessingActive" bit regardless of whether there was a success or failure.
+            // Otherwise, the processing would be stuck forever.
+            isProcessingActive.set(false);
+        }
+        Instant processingFinished = clock.instant();
+        logger.trace("Processing scheduled tasks finished, took {}ms", Duration.between(processingStarted, processingFinished).toMillis());
+    }
+
+    private void processScheduledTasksInternal() {
+        if (scheduledTasks.isEmpty()) {
+            // There are no scheduled tasks, hence, nothing to do
+            return;
+        }
+        long currentTimeMillis = clock.millis();
+        ScheduledTransformTask scheduledTask = scheduledTasks.first();
```

Comment:
am I getting this right? `processScheduledTasksInternal` only triggers 1 transform at a time? so if I have 100 transforms it takes 500s until all have been triggered?

Is it the intention to have this as a "rate limiter" ?
Nevertheless, if I have several transforms and want to schedule them every `10s`, they would get delayed already with 2 transforms, but I think we should allow it. We might break existing transforms with this.

Reply:
Done.
This is now fixed so the scheduler processes all the tasks until the current time.